Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,689 @@
+    switch self.state {
+    case .idle:
+      self.handleError(GRPCError.ProtocolViolation("Message received before headers"))
+    case let .active(requestStreamSource, _, _, _):
+      // TODO: Handle YieldResult from `yield()`
+      requestStreamSource.yield(request)
+    case .finishingSuccessfully:
+      self.handleError(GRPCError.ProtocolViolation("Message received after end of stream"))
+    case .completed:
+      // We received a message but we're already done: this may happen if we terminate the RPC
+      // due to a channel error, for example.
+      ()
+    }
+  }
+
+  @inlinable
+  internal func receiveInterceptedEnd() {
+    switch self.state {
+    case .idle:
+      self.handleError(GRPCError.ProtocolViolation("End of stream received before headers"))
+    case let .active(requestStreamSource, _, _, _):
+      // TODO: Handle YieldResult from `finish()`
+      requestStreamSource.finish()
+    case .finishingSuccessfully:
+      self.handleError(GRPCError.ProtocolViolation("Multiple end of stream blocks received on RPC"))
+    case .completed:
+      // We received a message but we're already done: this may happen if we terminate the RPC
+      // due to a channel error, for example.
+      ()
+    }
+  }
+
+  // MARK: - User Function To Interceptors
+
+  @inlinable
+  internal func interceptResponse(
+    _ response: Response,
+    metadata: MessageMetadata,
+    promise: EventLoopPromise<Void>?
+  ) {
+    self.context.eventLoop.assertInEventLoop()
+    switch self.state {
+    case .idle:
+      // The user handler cannot send responses before it has been invoked.
+      preconditionFailure()
+
+    case .active:
+      self.interceptors.send(.message(response, metadata), promise: promise)
+
+    case .finishingSuccessfully:
+      // TODO: Add test for this...
+      // The user handler cannot send responses once it has finished.
+      // ... or can it? (is this called from the delegate?)
+      preconditionFailure()
+
+    case .completed:
+      promise?.fail(GRPCError.AlreadyComplete())
+    }
+  }
+
+  @inlinable
+  internal func userHandlerCompleted(_ result: Result<GRPCStatus, Error>) {
+    switch self.state {
+    case .idle:
+      // The user handler cannot complete before it is invoked.
+      preconditionFailure()
+
+    case let .active(_, context, responseStreamWriter, _):
+      switch result {
+      case let .success(status):
+        // The user handler is done but there may be pending writes in the response stream writer.
+
+        // Create a promise to hang a callback off when the response stream is drained.
+        let responseStreamDrainedPromise: EventLoopPromise<GRPCStatus> = self.context.eventLoop
+          .makePromise()
+
+        // Register callback for the response stream being drained.
+        responseStreamDrainedPromise.futureResult.whenComplete(self.responseStreamDrained(_:))
+
+        self.responseStreamDrainTask = responseStreamDrainedPromise.completeWithTask {
```

Comment:
I actually had it this way and decided it was a race condition. Once the `userHandler` has returned we await the finish, we still have things coming in the handler. We want to have moved the state on. I tried moving the state on in this block but that meant a data race because the state was being moved in two different contexts.

Reply:
I've updated this now. Should be more what you were expecting to see.